File: src/main/java/qnd/KafkaProducerService.java
package qnd;

import java.util.Properties;

import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.common.serialization.StringSerializer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

/**
 * @author Vitali Henne (vitali.henne@freiheit.com).
 */
@Service
public class KafkaProducerService {
    private static final String TOPIC = "qnd";
    private static final Logger LOG = LoggerFactory.getLogger( KafkaProducerService.class );

    final private KafkaProducer<String, String> producer;

    public KafkaProducerService() {
        final Properties props = new Properties();
        props.setProperty( ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "kafka:9092" );
        props.setProperty( ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName() );
        props.setProperty( ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName() );
        props.setProperty( ProducerConfig.CLIENT_ID_CONFIG, "qnd-producer" );

        producer = new KafkaProducer<>( props );
    }

    public void sendMessageSynchronously( final String msg ) {
        // TODO implement me:
        // 1. create ProducerRecord with the TOPIC, key (optional), message
        // 2. send it with producer.send
        // 3. resolve the future
    }

    public void sendMessageAsynchronously( final String msg ) {
        // TODO implement me
        // 1. create ProducerRecord with the TOPIC, key (optional), message
        // 2. send it with producer.send
    }
}

File: src/main/java/qnd/KafkaProducerService.kt
package qnd

import mu.KotlinLogging
import org.apache.kafka.clients.producer.KafkaProducer
import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.clients.producer.ProducerRecord

class KafkaProducerService {
    private val logger = KotlinLogging.logger {}
    private val producer: KafkaProducer<String, String>

    init {
        val config = mapOf(
            ProducerConfig.BOOTSTRAP_SERVERS_CONFIG to Config.KAFKA_URL,
            ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG to Config.keySerializerClassName,
            ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG to Config.valueSerializerClassName,
            ProducerConfig.CLIENT_ID_CONFIG to "qnd-producer"
        )

        producer = KafkaProducer(config)

        logger.info { "created producer service" }
    }

    fun sendMessageSynchronously(msg: String) {
        // TODO implement me
        // 1. create ProducerRecord with the topic Config.TEST_TOPIC, key (optional), message
        // 2. send it with producer.send
        // 3. resolve the future
    }

    fun sendMessageAsynchronously(msg: String) {
        // TODO implement me
        // 1. create ProducerRecord with the Config.TEST_TOPIC, key (optional), message
        // 2. send it with producer.send
    }
}